File: collatz.c
#include <stdio.h>
#include <time.h>
struct collatz_struct
{
    int number;
    int cycle_length;
    int max;
} collatz(int number)
{

    int max = number;
    int init_number = number;
    int cycle_length = 0;
    struct collatz_struct result;
    result.number = number;
    while (number > 1)
    {
        if ((number & 1) == 0) // even
        {
            number = number >> 1;
        }
        else // odd
        {
            number = (number << 1) + number + 1;
        }
        cycle_length += 1;
        if (number > max)
        {
            max = number;
        }
    }

    result.cycle_length = cycle_length;
    result.max = max;

    return result;
}

#define MAX 10000000
int numbers[MAX];
int maxima[MAX];
int cycles[MAX];

int main()
{

    int i;
    clock_t start_time, end_time;
    double cpu_time_used;

    start_time = clock();

    // ... existing code ...

    for (i = 3; i < MAX; i += 2)
    {
        struct collatz_struct r = collatz(i);
        numbers[i] = r.number;
        cycles[i] = r.cycle_length;
        maxima[i] = r.max;
        if (i % 10000000 == 0)
        {
            printf("%d\n", i);
        }
        //  printf("%d %d %d\n", numbers[i], cycles[i], maxima[i]);
    }
    end_time = clock();
    cpu_time_used = ((double)(end_time - start_time)) / CLOCKS_PER_SEC;

    printf("Time taken: %f seconds\n", cpu_time_used);
    return 0;
}

File: collatz.py

import matplotlib.pyplot as plt
import time
import numba
import numpy as np

MAX = 100000
numbers = np.zeros(MAX, dtype=np.int32)
maxima = np.zeros(MAX, dtype=np.int32)
cycles = np.zeros(MAX, dtype=np.int32)

n = np.array


@numba.njit
def collatz(number):
    max = number
    init_number = number
    cycle_length = 1
    while number > 1:
        if not number & 1:
            number >>= 1
        else:
            number += (number << 1) + 1
        cycle_length += 1
        if number > max:
            max = number
    return cycle_length, max


collatz(2)

times_start = time.time()
for i in range(2, MAX):
    cycle, max = collatz(i)
    numbers[i] = i  # [i] = cycle
    cycles[i] = cycle   # [i] = cycle
    maxima[i] = max  # [i] = max
times_end = time.time()
print(times_end - times_start)

with open('collatz.txt', 'w') as f:
    for i in range(1, MAX):
        # {numbers[i]:>3},
        f.write(f"{i:>3}, {cycles[i]:>5} {maxima[i]:>7}\n")
plt.figure(figsize=(10, 10))
plt.subplot(2, 1, 1)

plt.plot(numbers, cycles)

plt.title('Number of Cycles')
plt.subplot(2, 1, 2)
plt.plot(numbers, maxima)
plt.title('Maximum value reached')
plt.show()
